compact/range.go

```diff
@@ -209,6 +231,48 @@ func (r *Range) appendImpl(end uint64, seed []byte, hashes [][]byte, visitor Vis
 	return nil
 }
 
+// intersectImpl returns the intersection of two compact ranges. It also
+// modifies the `r` and `other` compact ranges in such a way that they become
+// adjacent, and a subsequent AppendRange operation between them will result in
+// a compact range that represents the union of the two original ranges.
+func (r *Range) intersectImpl(other *Range) (*Range, error) {
+	if other.f != r.f {
+		return nil, errors.New("incompatible ranges")
+	} else if other.begin < r.begin {
+		return nil, errors.New("ranges unordered")
+	}
+
+	if other.end <= r.end { // The other range is nested.
+		intersection := *other             // Note: Force the clone.
+		*other = *r.f.NewEmptyRange(r.end) // Note: Force the rewrite.
+		return &intersection, nil
+	}
+
+	begin, end := other.begin, r.end
+	if begin > end { // The other range is disjoint.
+		return nil, fmt.Errorf("ranges are disjoint: other.begin=%d, want <= %d", begin, end)
+	} else if begin == end { // The ranges touch ends.
```

Comment:
I don't see the distinction here, since the intersection of two disjoint ranges is the empty set regardless of whether or not they "almost" intersect.